File: workspace-2025-1c/ModeloParcial/src/parciales/GestionVentas.java
package parciales;

import java.io.*;
import java.util.*;

public class GestionVentas {
	private List<Venta> ventas;

	public GestionVentas(String archivoVentas) {
		ventas = new ArrayList<>();
		cargarVentas(archivoVentas);
	}

	private void cargarVentas(String archivo) {
		try (BufferedReader br = new BufferedReader(new FileReader(archivo))) {
			String linea;
			while ((linea = br.readLine()) != null) {
				String[] partes = linea.split(",");
				if (partes.length != 4) {
					System.out.println("Línea mal formada: " + linea);
					continue;
				}
				String cliente = partes[0].trim();
				String producto = partes[1].trim();
				int cantidad = Integer.parseInt(partes[2].trim());
				double precio = Double.parseDouble(partes[3].trim());

				ventas.add(new Venta(cliente, producto, cantidad, precio));
			}
		} catch (IOException e) {
			System.out.println("Error leyendo el archivo: " + e.getMessage());
		} catch (NumberFormatException e) {
			System.out.println("Error en conversión de números: " + e.getMessage());
		}
	}

	public Map<String, List<Venta>> ventasPorCliente() {
		Map<String, List<Venta>> mapa = new HashMap<>();
		for (Venta v : ventas) {
			mapa.computeIfAbsent(v.getCliente(), k -> new ArrayList<>()).add(v);
		}
		return mapa;
	}

	public String clienteTop() {
		Map<String, Double> totales = new HashMap<>();
		for (Venta v : ventas) {
			totales.put(v.getCliente(), totales.getOrDefault(v.getCliente(), 0.0) + v.getMontoTotal());
		}
		return totales.entrySet().stream().max(Map.Entry.comparingByValue()).map(Map.Entry::getKey).orElse(null);
	}

	public String productoMasVendido() {
		Map<String, Integer> cantidades = new HashMap<>();
		for (Venta v : ventas) {
			cantidades.put(v.getProducto(), cantidades.getOrDefault(v.getProducto(), 0) + v.getCantidad());
		}
		return cantidades.entrySet().stream().max(Map.Entry.comparingByValue()).map(Map.Entry::getKey).orElse(null);
	}

	public static void main(String[] args) {
		GestionVentas gestion = new GestionVentas("ventas.csv");

		System.out.println("Ventas por cliente:");
		for (Map.Entry<String, List<Venta>> entry : gestion.ventasPorCliente().entrySet()) {
			System.out.print(entry.getKey() + ": ");
			for (Venta v : entry.getValue()) {
				System.out.print(v.getProducto() + " x" + v.getCantidad() + "  ");
			}
			System.out.println();
		}

		System.out.println("\nCliente top: " + gestion.clienteTop());
		System.out.println("Producto más vendido: " + gestion.productoMasVendido());
	}
}

File: 07-integradores/gestion_ventas.py
from io import open


class Venta:
    def __init__(self, cliente, producto, cantidad, precio_unitario):
        self.cliente = cliente
        self.producto = producto
        self.cantidad = cantidad
        self.precio_unitario = precio_unitario

    def monto_total(self):
        return self.cantidad * self.precio_unitario


class GestionVentas:
    def __init__(self, archivo_ventas):
        self.ventas = []
        self.cargar_ventas(archivo_ventas)

    def cargar_ventas(self, archivo):
        try:
            with open(archivo, "r", encoding="utf-8") as f:
                for linea in f:
                    partes = linea.strip().split(",")
                    if len(partes) != 4:
                        print(f"Línea mal formada: {linea}")
                        continue
                    cliente, producto, cantidad_str, precio_str = partes
                    try:
                        cantidad = int(cantidad_str)
                        precio = float(precio_str)
                        venta = Venta(
                            cliente.strip(), producto.strip(), cantidad, precio
                        )
                        self.ventas.append(venta)
                    except ValueError:
                        print(f"Error en conversión de números: {linea}")
        except FileNotFoundError:
            print(f"Archivo no encontrado: {archivo}")

    def ventas_por_cliente(self):
        resultado = {}
        for venta in self.ventas:
            resultado.setdefault(venta.cliente, []).append(venta)
        return resultado

    def cliente_top(self):
        totales = {}
        for venta in self.ventas:
            totales[venta.cliente] = totales.get(venta.cliente, 0) + venta.monto_total()
        if totales:
            return max(totales, key=totales.get)
        return None

    def producto_mas_vendido(self):
        contador = {}
        for venta in self.ventas:
            contador[venta.producto] = contador.get(venta.producto, 0) + venta.cantidad
        if contador:
            return max(contador, key=contador.get)
        return None


# Código de prueba


def main():
    gestion = GestionVentas("ventas.csv")
    print("Ventas por cliente:")
    for cliente, ventas in gestion.ventas_por_cliente().items():
        print(f"{cliente}: {[f'{v.producto} x{v.cantidad}' for v in ventas]}")

    print("\nCliente top:", gestion.cliente_top())
    print("Producto más vendido:", gestion.producto_mas_vendido())


if __name__ == "__main__":
    main()
